### ISSUE TITLE
Allow Plug.Conn.assigns setting from a LV

### ISSUE THREAD

**Issue by guidotripaldi** [phoenixframework/phoenix_live_view#288]
Currently it is not possible to manipulate the `@conn` assigns from inside a mounted LV, e.g. to control the contents of the layout in which the LV is enclosed.

This could instead to be very useful if, for example, it was necessary to highlight the correct tab in a navbar present in the header of the `app.html.eex` layout, based on that specific LV (e.g. setting a  @current_page assignment to the name of the current LV):

```eex
# templates/layout/app.html.eex
<!DOCTYPE html>
<html lang="en">
  ...
    <header>
      <%= render_navbar(@current_page, @disabled_pages) %>
    </header>

    <main role="main" class="container">
      <%= render @view_module, @view_template, assigns %>
    </main>
    <script type="text/javascript" src="<%= Routes.static_path(@conn, "/js/app.js") %>"></script>
  </body>
</html>
```

Actually this is possible only if the LV is rendered from a Controller, for example:

```ex
# controllers/my_controller.ex
defmodule MyAppWeb.MyController do
  use MyAppWeb, :controller

  def page_1(conn, _params) do
    conn
    |> Plug.Conn.assign(:current_page, "page_1")
    |> Plug.Conn.assign(:disabled_pages, ["page_4", "page_5"])
    |> live_render(MyApp.MyLive, session: %{})
  end
  ...
```

Instead it is not possible from a LV because the `@conn` chain isn't  available. So I've tried to add this functionality adding the new `conn_assigns/1` optional callback to LV, that can be used this way inside a LV module:

```ex
# live/my_live.ex
defmodule MyApp.MyLive do
  use Phoenix.LiveView

  def render(assigns) do
    ...
  end

  def mount(session, socket) do
    ...
  end

 # This optional callback receives in input the `conn.assigns` and its returns updates the @conn
  def conn_assigns(assigns) do
    assigns
    |> Map.put(:current_page, "page_1")
    |> Map.put(:disabled_pages, ["page_4", "page_5"])
  end

  ...

end
```

if you think this functionality could be useful I can submit a PR, try to take a look at its code [in this branch](https://github.com/guidotripaldi/phoenix_live_view/tree/conn_assigns)

**Comment by chrismccord** [comment]
If you need to update the layout, it should be a live view and not a static view. The issue with setting conn assigns from the LV is the conn isn't there after the HTTP request, so we can't handle it properly once connected. Thanks!

**Comment by guidotripaldi** [comment]
Thank you Chris.

> If you need to update the layout, it should be a live view and not a static view

if I had understood well, you mean that if we need to render a updatable navbar we have to put it in a master LV instead inside the default layout, and compose the rest of the page with LV children (ala SPA), but this way a) if a live_link is used somewhere in the page its content will replace all the contents including the navbar. And b) it means we cannot build a hybrid application (some pages by LV, some pages by Controllers, with a common navbar on the top), if I'm right.

> The issue with setting conn assigns from the LV is the conn isn't there after the HTTP request

the function I added is called from Phoenix.LiveView.Controller.live_render that, if I had studied well the code, is in the middle of the @conn chain, so it should not broke anything. (but of course you are the Master, so these are just my humbling notes)


 actually I think the way is implemented doesn't broke
